helix-contract/contracts/ln/base/LnBridgeTarget.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import "../interface/ILnBridgeSource.sol";
+import "./LnBridgeHelper.sol";
+
+contract LnBridgeTarget is LnBridgeHelper {
+    uint256 constant public MIN_REFUND_TIMESTAMP = 30 * 60;
+    struct TransferInfo {
+        uint48 nonce;
+        uint48 lastRefundNonce;
+        address slasher;
+    }
+
+    mapping(bytes32 => TransferInfo) public transferInfos;
+    mapping(bytes32 => address) public refundReceiver;
+
+    event TransferFilled(bytes32 transferId, address slasher);
+
+    // fill transfer
+    // 1. if transfer is not refund or relayed, LnProvider relay message to fill the transfer, and the transfer finished on target chain
+    // 2. if transfer is timeout and not processed, slasher(any account) can fill the transfer and request refund
+    // if it's filled by slasher, we store the address of the slasher
+    function _fillTransfer(
+        TransferParameter calldata params,
+        address slasher
+    ) internal returns(bytes32 transferId, uint48 lastRefundNonce) {
+        TransferInfo memory lastInfo = transferInfos[params.lastTransferId];
+        require(lastInfo.nonce + 1 == params.nonce, "Invalid last transferId");
+        transferId = keccak256(abi.encodePacked(
+            params.lastTransferId,
+            params.lastBlockHash,
+            params.nonce,
+            params.timestamp,
+            params.token,
+            params.receiver,
+            params.amount));
+        TransferInfo memory transferInfo = transferInfos[transferId];
+        require(transferInfo.nonce == 0, "lnBridgeTarget:message exist");
+        lastRefundNonce = lastInfo.slasher != address(0) ? params.nonce - 1 : lastInfo.lastRefundNonce;
```

Comment:
```suggestion
        // Find the previous refund fill, it is a refund fill if the slasher is not zero address.
        // We optimise storage gas by using nonce of fill id for that refund instead of directly use fill id.
        lastRefundNonce = lastInfo.slasher != address(0) ? lastInfo.nonce: lastInfo.lastRefundNonce;
```

Reply:
invalid latest slash transfer

Reply:
// 1. if previous.slasher != 0, then previous has been filled
 // 2. if previous.latestSlashTransferId != 0, then previous has been filled